lib/pure/net.nim

```diff
@@ -678,56 +672,49 @@ when defineSsl:
-    if pskString.len.cuint > max_psk_len:
-      return 0
-    if identityString.len.cuint >= max_identity_len:
-      return 0
-    copyMem(identity, identityString.cstring, identityString.len + 1) # with the last zero byte
-    copyMem(psk, pskString.cstring, pskString.len)
-
-    return pskString.len.cuint
-
-  proc `clientGetPskFunc=`*(ctx: SslContext, fun: SslClientGetPskFunc) =
-    ## Sets function that returns the client identity and the PSK based on identity
-    ## hint from the server.
-    ##
-    ## Only used in PSK ciphersuites.
-    ctx.getExtraInternal().clientGetPskFunc = fun
-    ctx.context.SSL_CTX_set_psk_client_callback(
-        if fun == nil: nil else: pskClientCallback)
-
-  proc serverGetPskFunc*(ctx: SslContext): SslServerGetPskFunc =
-    return ctx.getExtraInternal().serverGetPskFunc
+  template genpskServerCallback(pskfunc: SslServerGetPskFunc): auto =
+    proc pskServerCallback(ssl: SslCtx; identity: cstring; psk: ptr cuchar;
+        max_psk_len: cint): cuint {.cdecl.} =
+      let pskString = pskfunc($identity)
+      if pskString.len.cint > max_psk_len:
+        return 0
+      copyMem(psk, pskString.cstring, pskString.len)
+      return pskString.len.cuint
 
-  proc pskServerCallback(ssl: SslCtx; identity: cstring; psk: ptr cuchar;
-      max_psk_len: cint): cuint {.cdecl.} =
-    let ctx = SslContext(context: ssl.SSL_get_SSL_CTX)
-    let pskString = (ctx.serverGetPskFunc)($identity)
-    if pskString.len.cint > max_psk_len:
-      return 0
-    copyMem(psk, pskString.cstring, pskString.len)
+    pskServerCallback
 
-    return pskString.len.cuint
-
-  proc `serverGetPskFunc=`*(ctx: SslContext, fun: SslServerGetPskFunc) =
+  proc `serverGetPskFunc=`*(ctx: SslContext, fun: static SslServerGetPskFunc) =
     ## Sets function that returns PSK based on the client identity.
     ##
     ## Only used in PSK ciphersuites.
-    ctx.getExtraInternal().serverGetPskFunc = fun
-    ctx.context.SSL_CTX_set_psk_server_callback(if fun == nil: nil
-                                                else: pskServerCallback)
+    if not fun.isNil:
+      ctx.context.SSL_CTX_set_psk_server_callback(genpskServerCallback(fun))
+
+  template genpskClientCallback(pskfunc: SslClientGetPskFunc): auto =
+      proc pskClientCallback(ssl: SslPtr; hint: cstring; identity: cstring;
+          max_identity_len: cuint; psk: ptr cuchar;
+          max_psk_len: cuint): cuint {.cdecl.} =
+        let (identityString, pskString) = pskfunc($hint)
+        if pskString.len.cuint > max_psk_len:
+          return 0
+        if identityString.len.cuint >= max_identity_len:
+          return 0
+        copyMem(identity, identityString.cstring, identityString.len + 1) # with the last zero byte
+        copyMem(psk, pskString.cstring, pskString.len)
+        return pskString.len.cuint
+
+      pskClientCallback
+
+  proc `clientGetPskFunc=`*(ctx: SslContext, fun: static SslClientGetPskFunc) =
```

Comment:
Why `static` here?

Reply:
without `static`, the fun makes the callback into a closure, which is of course incompatible with openssl.

This is the tradeoff to make psk work at all: the callback function must be known at compile-time. As the test shows, this is not a big issue, ergonomically it's pretty transparent and even lambdas work (as long as they are .nimcall)

As I learned from following your work, accessing the SslClientGetPskFunc at runtime consistently isn't possible, as the callback only has access to the raw ssl context, not the Nim wrapper context.  Which is why your solution was to include it within the ex_data.  But to get it back out, the callback needs to know the index of said data. which it also cant know.

  Furthermore, we can't just assume that the index is 0, as the ex_data is not per context, but global across all contexts, so that solution doesn't work reliably, either.

So I concluded that to keep the abstraction of SslClientGetPskFunc it would have to be a compile-time rewriting, hence the template.

Reply:
a closure in nim is just 2 function pointers; doesn't openssl allow passing some sort of void* context, with which we could pass the environment needed for the closure?

Reply:
the callback doesn't allow for extra arguments, or we'd pass in the callback there. it's:
```nim
pskClientCallback(ssl: SslPtr; hint: cstring; identity: cstring; 
  max_identity_len: cuint; psk: ptr cuchar; max_psk_len: cuint): cuint
```

Reply:
perhaps the signature could be better expressed as `identity: var UncheckedArray[cchar]` and `psk: var UncheckedArray[cuchar]` but that's maybe yak shaving

Reply:
Since you already annotated `SslClientGetPskFunc` with `{.nimcall.}`, `static` shouldn't be necessary afaik.

Reply:
you'd have thought so, but if i remove the static i either get an error message saying something about :anonymous, or, with the current test, i just get a SIGSEGV. maybe that's a bug i should submit.

Reply:
> the sslcallback doesn't allow for extra arguments

here's how to pass callbacks: https://github.com/nim-lang/Nim/pull/17741#issuecomment-825933893